File: sdk/python/src/graft/transports/serial.py
"""GRAFT Serial Transport — pyserial-asyncio"""

from __future__ import annotations

from typing import Callable

try:
    import serial_asyncio
except ImportError:
    serial_asyncio = None  # type: ignore[assignment]


class SerialTransport:
    """Async serial transport using pyserial-asyncio."""

    def __init__(self, port: str, baudrate: int = 115200) -> None:
        if serial_asyncio is None:
            raise ImportError("pyserial-asyncio is required: pip install graft-py[serial]")
        self._port = port
        self._baudrate = baudrate
        self._reader: object | None = None
        self._writer: object | None = None
        self._connected = False
        self._handler: Callable[[bytes], None] | None = None
        self._task: object | None = None
        self.needs_cobs = True

    @property
    def connected(self) -> bool:
        return self._connected

    async def connect(self) -> None:
        import asyncio
        self._reader, self._writer = await serial_asyncio.open_serial_connection(
            url=self._port, baudrate=self._baudrate,
        )
        self._connected = True
        self._task = asyncio.create_task(self._read_loop())

    async def disconnect(self) -> None:
        self._connected = False
        if self._writer:
            self._writer.close()
        self._reader = None
        self._writer = None

    async def send(self, packet: bytes) -> None:
        if not self._connected or not self._writer:
            raise RuntimeError("SerialTransport: not connected")
        self._writer.write(packet)
        await self._writer.drain()

    def on_receive(self, handler: Callable[[bytes], None]) -> None:
        self._handler = handler

    async def _read_loop(self) -> None:
        buf = bytearray()
        while self._connected and self._reader:
            try:
                data = await self._reader.read(1024)
                if not data:
                    break
                buf.extend(data)
                while b"\x00" in buf:
                    idx = buf.index(b"\x00")
                    frame = bytes(buf[:idx])
                    buf = buf[idx + 1:]
                    if frame and self._handler:
                        from graft.core.cobs import cobs_decode
                        decoded = cobs_decode(frame)
                        if decoded is not None:
                            self._handler(decoded)
            except Exception:
                break
        self._connected = False

File: sdk/js/src/transports/serial.ts
/**
 * GRAFT Serial Transport — Node.js
 *
 * Uses the `serialport` npm package. COBS-framed with 0x00 delimiter.
 * Install: npm install serialport
 */

import type { GraftTransport } from './transport.js'

export interface SerialTransportOptions {
  /** Serial port path (e.g., '/dev/ttyUSB0', 'COM3') */
  path: string
  /** Baud rate (default: 115200) */
  baudRate?: number
}

export class SerialTransport implements GraftTransport {
  private _options: Required<SerialTransportOptions>
  private _port: any = null
  private _handler: ((packet: Uint8Array) => void) | null = null
  private _connected = false

  readonly needsCOBS = true

  constructor(options: SerialTransportOptions) {
    this._options = {
      path: options.path,
      baudRate: options.baudRate ?? 115200,
    }
  }

  get connected(): boolean {
    return this._connected
  }

  async connect(): Promise<void> {
    const { SerialPort } = await import('serialport')

    this._port = new SerialPort({
      path: this._options.path,
      baudRate: this._options.baudRate,
      autoOpen: false,
    })

    return new Promise((resolve, reject) => {
      this._port.open((err: Error | null) => {
        if (err) return reject(err)
        this._connected = true

        this._port.on('data', (chunk: Buffer) => {
          if (this._handler) {
            this._handler(new Uint8Array(chunk))
          }
        })

        this._port.on('close', () => {
          this._connected = false
        })

        this._port.on('error', () => {
          this._connected = false
        })

        resolve()
      })
    })
  }

  async disconnect(): Promise<void> {
    if (this._port && this._port.isOpen) {
      return new Promise((resolve) => {
        this._port.close(() => {
          this._connected = false
          resolve()
        })
      })
    }
    this._connected = false
  }

  async send(packet: Uint8Array): Promise<void> {
    if (!this._port || !this._connected) throw new Error('Serial: not connected')
    return new Promise((resolve, reject) => {
      this._port.write(Buffer.from(packet), (err: Error | null) => {
        if (err) reject(err)
        else resolve()
      })
    })
  }

  onReceive(handler: (packet: Uint8Array) => void): void {
    this._handler = handler
  }
}
